packages/dds/tree/src/util/utils.ts

```diff
@@ -4,13 +4,34 @@
  */
 
 import structuredClone from "@ungap/structured-clone";
+import { Brand, Opaque } from "./brand";
 
 /**
  * Make all transitive properties in T readonly
  */
-export type RecursiveReadonly<T> = {
-    readonly [P in keyof T]: RecursiveReadonly<T[P]>;
-};
+export type RecursiveReadonly<T> = T extends
+    | undefined
+    // eslint-disable-next-line @rushstack/no-new-null
+    | null
+    | boolean
+    | string
+    | number
+    // eslint-disable-next-line @typescript-eslint/ban-types
+    | Function
+    // eslint-disable-next-line @typescript-eslint/ban-types
+    | Symbol
+    ? T
+    : T extends Opaque<Brand<any, string>>
+    ? T
+    : T extends Map<infer K, infer V>
+    ? ReadonlyMap<RecursiveReadonly<K>, RecursiveReadonly<V>>
+    : T extends Set<infer K>
+    ? ReadonlySet<RecursiveReadonly<K>>
+    : T extends (infer V)[]
+    ? readonly RecursiveReadonly<V>[]
+    : {
+          readonly [K in keyof T]: RecursiveReadonly<T[K]>;
+      };
```

Comment:
I'd like to see tests for this. Ex:
```
// Type tests for RecursiveReadonly
{
    type _basic = requireTrue<areSafelyAssignable<RecursiveReadonly<number>, number>>;
    type _union = requireTrue<areSafelyAssignable<RecursiveReadonly<number | string>, number | string>>;
    type TestInner = Set<number>;
    type TestInnerReadonly = ReadonlySet<number>;
    type _set = requireTrue<areSafelyAssignable<RecursiveReadonly<Set<number>>, ReadonlySet<number>>>;
    type _setNested = requireTrue<areSafelyAssignable<RecursiveReadonly<Set<TestInnerReadonly>>, ReadonlySet<TestInnerReadonly>>>;
    type _map = requireTrue<areSafelyAssignable<RecursiveReadonly<Map<number, string>>, ReadonlyMap<number, string>>>;
    type _mapNestedKey = requireTrue<areSafelyAssignable<RecursiveReadonly<Map<TestInner, string>>, ReadonlyMap<TestInnerReadonly, string>>>;
    type _mapNestedValue = requireTrue<areSafelyAssignable<RecursiveReadonly<Map<string, TestInner>>, ReadonlyMap<string, TestInnerReadonly>>>;
}
```
SHoudl also cover the object case, and ensure unknown methods are stripped off the objects.